### PR TITLE
fix(artifacts): consider requiredArtifactIds in expected artifacts when trigger is pipeline type

### PR DESCRIPTION
**Pull request by nemesisOsorio**
fixes https://github.com/spinnaker/spinnaker/issues/6850

Expected artifacts can be used in automated triggers and stages,
as in the following example:

artifacts

if we trigger the previous pipeline with a pipeline stage like the next image

trigger-pipeline-using-stage

you will get the following error even if the parent pipeline provides the artifact

trigger-pipeline

in the example, I'm using a `Find Artifacts From Resource (Manifest)` stage 
and producing a docker image as an artifact, I believe this is a similar case for building
the docker image using Jenkins like is mentioned in https://github.com/spinnaker/spinnaker/issues/6850

Before the following PRs (i.e. in spinnaker < 1.30) `Deploy (Manifest)` stage was working
https://github.com/spinnaker/orca/pull/4322
https://github.com/spinnaker/orca/pull/4322
https://github.com/spinnaker/orca/pull/4322

the intent of this PR is to restore the previous behavior

Here are the pipelines

[pr-pipelines.zip](https://github.com/spinnaker/orca/files/12115106/pr-pipelines.zip)

### CODE REVIEW

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy [hunk 1]
```diff
@@ -93,7 +98,7 @@ class DependentPipelineStarter implements ApplicationContextAware {
       parameters           : [:],
       strategy             : suppliedParameters.strategy == true,
       correlationId        : "${parentPipeline.id}_${parentPipelineStageId}_${pipelineConfig.id}_${parentPipeline.startTime}".toString(),
-      expectedArtifactIds  : expectedArtifactIds
+      expectedArtifactIds  : expectedArtifactIds.toSet().toList()
```
Comment (dbyron-sf) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:101]: Why is this necessary?

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy [hunk 2]
```diff
@@ -85,6 +85,11 @@ class DependentPipelineStarter implements ApplicationContextAware {
       it.expectedArtifactIds ?: []
     }
 
+    def requiredArtifactIds = pipelineConfig.get("stages", []).collectMany {
```
Comment (dbyron-sf) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:88]: Please add a comment explaining what this does / why it's neceessary.

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy (hunk 1 above)
Comment (nemesisOsorio) [reply orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy:101]: only to avoid duplications between trigger and stages

**Comment by dbyron-sf** [comment]
It would also help to connect the dots to show if this ever worked in previous versions of spinnaker, what PR broke it, and ideally why.

**Comment by nemesisOsorio** [comment]
> It would also help to connect the dots to show if this ever worked in previous versions of spinnaker, what PR broke it, and ideally why.

Absolutely, I tested it out in `1.28` and `1.27` and it was working
There is a chance the intention for `expectedArtifacts` was never to use them in stages like `Deploy (Manifest)`, but right now, you can select an `expectedArtifacts` in the UI.

These are the related PRs
https://github.com/spinnaker/orca/pull/4322
https://github.com/spinnaker/orca/pull/4397
https://github.com/spinnaker/orca/pull/4404

**Comment by dbyron-sf** [comment]
> > It would also help to connect the dots to show if this ever worked in previous versions of spinnaker, what PR broke it, and ideally why.
> 
> Absolutely, I tested it out in `1.28` and `1.27` and it was working There is a chance the intention for `expectedArtifacts` was never to use them in stages like `Deploy (Manifest)`, but right now, you can select an `expectedArtifacts` in the UI.
> 
> These are the related PRs #4322 #4397 #4404

Great.  Those PRs look like they went in to 1.30, so I assume it's been broken since then.  Any idea about 1.29?

**Comment by dbyron-sf** [comment]
Also, cc: @jervi since you made those PRs.

**Comment by nemesisOsorio** [comment]
> > > It would also help to connect the dots to show if this ever worked in previous versions of spinnaker, what PR broke it, and ideally why.
> > 
> > 
> > Absolutely, I tested it out in `1.28` and `1.27` and it was working There is a chance the intention for `expectedArtifacts` was never to use them in stages like `Deploy (Manifest)`, but right now, you can select an `expectedArtifacts` in the UI.
> > These are the related PRs #4322 #4397 #4404
> 
> Great. Those PRs look like they went in to 1.30, so I assume it's been broken since then. Any idea about 1.29?

I tested it out in orca `release-1.29.x` and it is working, so you are right it's broken since `1.30`

**Review by edgarulg — APPROVED** [review]
Overall these changes looks good to me. This is an issue present in 1.30 & 1.31 only. I left my approve but I want @dbyron-sf or @xibz to have last word on approve these changes.

I also added @jervi as reviewer because these seems to be related with his PRs.

**Comment by edgarulg** [comment]
Also I'm seeing the build failing so would be good to rerun it because seems to be a false-positive error in the tests.

**Comment by dbyron-sf** [comment]
@nemesisOsorio I'm still hoping for a bit more in the description I think. Something like
```
Before PR XYZ (i.e. in spinnaker < 1.30), when pipeline B contained a trigger on
the completion of pipeline A with an artifact constraint, pipeline B only
triggered when pipeline A provided that artifact via the XXX field in the
pipeline execution context.  If pipeline A didn't provide that artifact,
pipeline B wouldn't trigger.  After that PR, pipeline B wouldn't trigger either
way -- even if pipeline A provided that artifact.  This PR restores the pre-PR
XYZ behavior.
```
but with the important missing bits filled in.

**Review by jervi — APPROVED** [review]
Sorry for overlooking this artifact flow when creating the aforementioned PR's 😬
But I think this is a good fix @nemesisOsorio.

**Comment by xibz** [comment]
@nemesisOsorio Thank you for the changes and updates! This looks a lot better :).

> Sorry for overlooking this artifact flow
@jervi No worries! It happens. Im trying to think of any major features I've written that contained 0 bugs. Yep, can't think of one haha.

**Review by dbyron-sf — APPROVED** [review]

**Comment by dbyron-sf** [comment]
@mergifyio backport release-1.30.x release-1.31.x
